src/deploy.ts

```diff
@@ -190,14 +203,128 @@ class Deployer {
+            ).stdout;
+            await this.apiClient.postProjectEnvironment(deployTarget.project.id, {
+              source: {
+                provider: authedRepo.provider,
+                provider_id: authedRepo.provider_id,
+                url: authedRepo.url,
+                branch
+              }
+            });
+            return true;
+          } else {
+            // repo not auth’ed; link to auth page and poll for auth
+            // TODO: link to internal page that bookends the flow and handles the no-oauth-token case more gracefully
+            this.effects.clack.log.info(
+              `Authorize Observable to access the ${bold(repoName)} repository: ${link(
+                "https://github.com/apps/observable-data-apps-dev/installations/select_target"
+              )}`
+            );
+            const spinner = this.effects.clack.spinner();
+            spinner.start("Waiting for repository to be authorized");
+            const pollExpiration = Date.now() + DEPLOY_POLL_MAX_MS;
+            while (true) {
+              if (Date.now() > pollExpiration) {
+                spinner.stop("Waiting for repository to be authorized timed out.");
+                throw new CliError("Repository authorization failed");
+              }
+              const repositories = (await this.apiClient.getGitHubRepositories())?.repositories;
+              const authedRepo = repositories?.find(({url}) => formatGitUrl(url) === repoName);
+              if (authedRepo) {
+                spinner.stop("Repository authorized.");
+                await this.apiClient.postProjectEnvironment(deployTarget.project.id, {
+                  source: {
+                    provider: authedRepo.provider,
+                    provider_id: authedRepo.provider_id,
+                    url: authedRepo.url,
+                    branch: null // TODO detect branch
+                  }
+                });
+                return true;
+              }
+              await new Promise((resolve) => setTimeout(resolve, 2000));
+            }
+          }
+        } else {
+          this.effects.clack.log.error("No GitHub remote found; cannot enable continuous deployment.");
+        }
+      } else {
+        this.effects.clack.log.error("Not at root of a git repository; cannot enable continuous deployment.");
+      }
+    }
+    return false;
+  }
+
+  private async startNewDeploy(): Promise<GetDeployResponse> {
+    const {deployConfig, deployTarget} = await this.getDeployTarget(await this.getUpdatedDeployConfig());
+    let deployId: string | null;
+    if (deployConfig.continuousDeployment) {
+      deployId = await this.cloudBuild(deployTarget);
+    } else {
+      const buildFilePaths = await this.getBuildFilePaths();
+      deployId = await this.createNewDeploy(deployTarget);
+      await this.uploadFiles(deployId, buildFilePaths);
+      await this.markDeployUploaded(deployId);
+    }
     return await this.pollForProcessingCompletion(deployId);
```

Comment:
this pollForProcessingCompletion pattern, in the continuous deployment case, keeps the user's terminal busy even though at that point all the work is being done on our servers. i mean, they can always ctrl-c, and we could tell them it's safe to do that. we could return early and say "when your deploy is done, if it succeeds, it will be visible at this url". but maybe it's better to wait so that we can return an error code if it fails?

Reply:
fil thinks it's correct as-is; if someone has scripted this, they can run `yarn deploy` to kick off a cloud build, and they will receive a CliError code if the cloud deploy fails.

Reply:
Given a magic wand, would you want to stream the build logs here?

Reply:
@mythmon Eventually, yeah!

Reply:
sticking with this pattern for now